File: backend/routes/realtime.js
const express = require('express');
const db = require('../config/database');
const dataService = require('../services/dataService');

const router = express.Router();

// Fetch real-time station data
router.get('/stations/live', async (req, res) => {
  try {
    const stations = dataService.getStations();
    const formattedStations = stations.map(s => ({
      station_code: s.stnCode,
      station_name: s.stnName,
      city: s.stnCity,
      state: s.stnCity // Using city as state since state is not in JSON
    }));
    
    res.json({
      success: true,
      count: formattedStations.length,
      data: formattedStations
    });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch stations' });
  }
});

// Fetch real-time train data
router.get('/trains/live', async (req, res) => {
  try {
    const trains = dataService.getTrains();
    const formattedTrains = trains.map(t => ({
      train_number: t.properties.number,
      train_name: t.properties.name,
      train_type: t.properties.type
    }));
    
    res.json({
      success: true,
      count: formattedTrains.length,
      data: formattedTrains
    });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch trains' });
  }
});

// Fetch real-time train schedule
router.get('/trains/:trainNumber/live-schedule', async (req, res) => {
  const { trainNumber } = req.params;
  
  try {
    const train = dataService.findTrainByNumber(trainNumber);
    
    if (!train) {
      return res.status(404).json({ error: 'Train not found' });
    }

    // Extract coordinates and create schedule
    const coordinates = train.geometry.coordinates;
    const schedule = coordinates.map((coord, index) => ({
      stop_number: index + 1,
      station_code: index === 0 ? train.properties.from_station_code : 
                    index === coordinates.length - 1 ? train.properties.to_station_code : 
                    `STN${index}`,
      station_name: index === 0 ? train.properties.from_station_name : 
                    index === coordinates.length - 1 ? train.properties.to_station_name : 
                    `Station ${index}`,
      arrival_time: index === 0 ? null : train.properties.departure,
      departure_time: index === coordinates.length - 1 ? null : train.properties.arrival,
      distance_km: Math.round((train.properties.distance / coordinates.length) * index),
      day_offset: 0
    }));
    
    res.json({
      success: true,
      trainNumber,
      stops: schedule.length,
      data: schedule
    });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch schedule' });
  }
});

// Real-time seat availability
router.get('/trains/:trainNumber/availability', async (req, res) => {
  const { trainNumber } = req.params;
  const { date, class: classType } = req.query;
  
  try {
    const train = dataService.findTrainByNumber(trainNumber);
    
    if (!train) {
      return res.status(404).json({ error: 'Train not found' });
    }

    // Generate mock availability data based on train classes
    const classes = ['SL', '3A', '2A', '1A'];
    const availability = classes.map(cls => ({
      class_type: cls,
      available_seats: Math.floor(Math.random() * 50) + 10,
      total_seats: cls === 'SL' ? 72 : cls === '3A' ? 64 : cls === '2A' ? 48 : 24,
      base_fare: cls === 'SL' ? 300 : cls === '3A' ? 800 : cls === '2A' ? 1500 : 2500,
      train_name: train.properties.name
    }));

    const filteredAvailability = classType 
      ? availability.filter(a => a.class_type === classType)
      : availability;
    
    res.json({
      success: true,
      trainNumber,
      date,
      data: filteredAvailability
    });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch availability' });
  }
});

module.exports = router;

File: backend/routes/realtime.py
import random
from flask import Blueprint, request, jsonify
from services.data_service import get_stations, get_trains, find_train_by_number

realtime_bp = Blueprint('realtime', __name__)


@realtime_bp.route('/stations/live', methods=['GET'])
def stations_live():
    stations = get_stations()
    formatted = [
        {"station_code": s.get('stnCode'), "station_name": s.get('stnName'), "city": s.get('stnCity'), "state": s.get('stnState', s.get('stnCity'))}
        for s in stations
    ]
    return jsonify({"success": True, "count": len(formatted), "data": formatted})


@realtime_bp.route('/trains/live', methods=['GET'])
def trains_live():
    trains = get_trains()
    formatted = [
        {"train_number": t['properties'].get('number'), "train_name": t['properties'].get('name'), "train_type": t['properties'].get('type')}
        for t in trains
    ]
    return jsonify({"success": True, "count": len(formatted), "data": formatted})


@realtime_bp.route('/trains/<train_number>/live-schedule', methods=['GET'])
def live_schedule(train_number):
    train = find_train_by_number(train_number)
    if not train:
        return jsonify({"error": "Train not found"}), 404

    coordinates = train.get('geometry', {}).get('coordinates', [])
    total = len(coordinates)
    distance = train['properties'].get('distance', 0)

    schedule = []
    for i, coord in enumerate(coordinates):
        schedule.append({
            "stop_number": i + 1,
            "station_code": train['properties'].get('from_station_code') if i == 0
                            else train['properties'].get('to_station_code') if i == total - 1
                            else f"STN{i}",
            "station_name": train['properties'].get('from_station_name') if i == 0
                            else train['properties'].get('to_station_name') if i == total - 1
                            else f"Station {i}",
            "arrival_time": None if i == 0 else train['properties'].get('departure'),
            "departure_time": None if i == total - 1 else train['properties'].get('arrival'),
            "distance_km": round((distance / total) * i) if total else 0,
            "day_offset": 0,
        })

    return jsonify({"success": True, "trainNumber": train_number, "stops": len(schedule), "data": schedule})


@realtime_bp.route('/trains/<train_number>/availability', methods=['GET'])
def availability(train_number):
    class_type = request.args.get('class')
    train = find_train_by_number(train_number)
    if not train:
        return jsonify({"error": "Train not found"}), 404

    classes = [
        {"class_type": "SL", "total_seats": 72, "base_fare": 300},
        {"class_type": "3A", "total_seats": 64, "base_fare": 800},
        {"class_type": "2A", "total_seats": 48, "base_fare": 1500},
        {"class_type": "1A", "total_seats": 24, "base_fare": 2500},
    ]
    result = [
        {**c, "available_seats": random.randint(10, c['total_seats']), "train_name": train['properties'].get('name')}
        for c in classes
    ]
    if class_type:
        result = [r for r in result if r['class_type'] == class_type]

    return jsonify({"success": True, "trainNumber": train_number, "data": result})
